src/token-processor/stacks-node/stacks-node-rpc-client.ts

```diff
@@ -0,0 +1,133 @@
+import { request } from 'undici';
+import { ClarityType, ClarityValue, cvToHex, hexToCV, UIntCV } from '@stacks/transactions';
+import { ENV } from '../../env';
+import { RetryableJobError } from '../queue/errors';
+
+interface ReadOnlyContractCallSuccessResponse {
+  okay: true;
+  result: string;
+}
+
+interface ReadOnlyContractCallFailResponse {
+  okay: false;
+  cause: string;
+}
+
+export type ReadOnlyContractCallResponse =
+  | ReadOnlyContractCallSuccessResponse
+  | ReadOnlyContractCallFailResponse;
+
+/**
+ * Performs read-only contract calls against a configured Stacks node. Performs type checking and
+ * returns data as decoded Clarity values.
+ */
+export class StacksNodeRpcClient {
+  private readonly contractAddress: string;
+  private readonly contractName: string;
+  private readonly senderAddress: string;
+  private readonly basePath: string;
+
+  constructor(args: { contractPrincipal: string; senderAddress: string }) {
+    [this.contractAddress, this.contractName] = args.contractPrincipal.split('.');
+    this.senderAddress = args.senderAddress;
+    this.basePath = `http://${ENV.STACKS_NODE_RPC_HOST}:${ENV.STACKS_NODE_RPC_PORT}`;
+  }
+
+  async readStringFromContract(
+    functionName: string,
+    functionArgs: ClarityValue[] = []
+  ): Promise<string | undefined> {
+    const clarityValue = await this.makeReadOnlyContractCall(functionName, functionArgs);
+    return this.checkAndParseString(clarityValue);
+  }
+
+  async readUIntFromContract(
+    functionName: string,
+    functionArgs: ClarityValue[] = []
+  ): Promise<bigint | undefined> {
+    const clarityValue = await this.makeReadOnlyContractCall(functionName, functionArgs);
+    const uintVal = this.checkAndParseUintCV(clarityValue);
+    try {
+      return BigInt(uintVal.value.toString());
+    } catch (error) {
+      throw new RetryableJobError(`Invalid uint value '${uintVal.value}'`);
+    }
+  }
+
+  private async sendReadOnlyContractCall(
+    functionName: string,
+    functionArgs: ClarityValue[]
+  ): Promise<ReadOnlyContractCallResponse> {
+    const body = {
+      sender: this.senderAddress,
+      arguments: functionArgs.map(arg => cvToHex(arg)),
+    };
+    const result = await request(
+      `${this.basePath}/v2/contracts/call-read/${this.contractAddress}/${this.contractName}/${functionName}`,
+      {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(body),
+      }
+    );
+    return (await result.body.json()) as ReadOnlyContractCallResponse;
```

Comment:
I've often found these error objects lacking in good debug information, and usually try/catch here to include things like the request URL, http response status code, etc in the error message.

Sometimes I've also used the approach of first grabbing the response as a string e.g. `result.body.text()`, in order to include that in the error if it fails to parse as JSON (you could imagine a non-JSON error response, like from a reverse-proxy).

Reply:
Done